apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/WatcherInterceptor.kt

```diff
@@ -8,59 +8,40 @@ import com.apollographql.apollo3.api.Query
 import com.apollographql.apollo3.cache.normalized.ApolloStore
 import com.apollographql.apollo3.cache.normalized.api.dependentKeys
 import com.apollographql.apollo3.cache.normalized.isRefetching
-import com.apollographql.apollo3.cache.normalized.watch
-import com.apollographql.apollo3.exception.ApolloException
+import com.apollographql.apollo3.cache.normalized.watchContext
 import com.apollographql.apollo3.interceptor.ApolloInterceptor
 import com.apollographql.apollo3.interceptor.ApolloInterceptorChain
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.catch
 import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.flow.emitAll
 import kotlinx.coroutines.flow.filter
 import kotlinx.coroutines.flow.flow
 import kotlinx.coroutines.flow.map
-import kotlinx.coroutines.flow.onCompletion
 import kotlinx.coroutines.flow.onEach
-import kotlinx.coroutines.flow.onStart
 
 internal class WatcherInterceptor(val store: ApolloStore) : ApolloInterceptor {
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
-    if (!request.watch) {
-      return chain.proceed(request)
-    }
+    val watchContext = request.watchContext ?: return chain.proceed(request)
 
     check(request.operation is Query) {
       "It's impossible to watch a mutation or subscription"
     }
 
     val customScalarAdapters = request.executionContext[CustomScalarAdapters]!!
-    var watchedKeys: Set<String>? = null
-    var isRefetching = false
+
+    @Suppress("UNCHECKED_CAST")
+    var watchedKeys: Set<String>? = watchContext.data?.let { store.normalize(request.operation, it as D, customScalarAdapters).values.dependentKeys() }
 
     return store.changedKeys
-        .onStart {
-          // Trigger the initial fetch
-          emit(emptySet())
-        }
         .filter { changedKeys ->
           watchedKeys == null || changedKeys.intersect(watchedKeys!!).isNotEmpty()
         }.map {
-          chain.proceed(request.newBuilder().isRefetching(isRefetching).build())
-              .catch {
-                if (it !is ApolloException) {
-                  // Re-throw cancellation exceptions
-                  throw it
-                }
-                // Else just ignore errors
-              }
+          chain.proceed(request.newBuilder().isRefetching(true).build())
```

Comment:
I was thinking we could even remove `isRefetching` completely so that `refetchPolicy` is only used by the `watch(fetchError, refetchError)` function? "refetching" would be a concept only known by this one and have no effect else? Keep things more local?

Reply:
The thing is currently, `isRefetching` is used to route to either `fetchPolicyInterceptor` or `refetchPolicyInterceptor`. And since users can set their own  `refetchPolicyInterceptor` we want to keep this mechanism? Or was this never released so we can remove it?

Reply:
Also, passing `fetchError` and `refetchError` kind of breaks the usual pattern of how `ApolloCalls` are configured. Maybe this instead?

```kotlin
apolloClient.query(query)
             .fetchPolicy(CacheFirst)
             .refetchPolicy(CacheOnly)
             .errorHandling(THROW_CACHE_AND_NETWORK_ERRORS)
             .refetchErrorHandling(IGNORE_ERROR)
             .watch()
``
`

2 advantages: 

1. It's more consistent
2. It allows to clone the call and keep all parameters (not sure if that is ever going to be used but maybe? )

Reply:
(had another comment but deleted it to keep this conversation focused)

We definitely want to keep the behaviour but the choice can be made by `watch(errorHandling, refetchErrorHandling)`:

```kotlin
fun <D : Query.Data> ApolloCall<D>.watch(
    fetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
    refetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
): Flow<ApolloResponse<D>> {
  var data: D? = null
  return toFlow()
      .catch {
        maybeThrow(it, fetchErrorHandling)
      }.onEach {
        data = it.data
      }.onCompletion {
        emitAll(
            copy().fetchPolicyInterceptor(refetchPolicyInterceptor).watch(data)
                .catch {
                  maybeThrow(it, refetchErrorHandling)
                }
        )
      }
}
```

Reply:
Oooh! Indeed, didn't think of that :). I'll make the change 👍